### DISCUSSION TITLE
Derived classes from json class

### DISCUSSION THREAD

**Discussion by bradacsa** [nlohmann/json#4313] [Q&A]
I just wrote two custom classes based on nlohmann::json:
`class Cookie : public nlohmann::json` <- this can handle cookies, chops up cookie strings from headers, etc...
`class JsonFile : public nlohmann::json` <- this is nothing else but the json class joined up with a fstream to make easier to load and save json from file.
My problem is, if I create a Cookie like
`Cookie A1("A1=d=AQABBN3h6GUCECMYqmLClG1pK2uiuaCh0xoFEgABCAEq6mUTZutMb2UBAiAAAAcI3OHoZXSoivI&S=AQAAAqvfHaGFJkLfQ07HNDr2tH0;Expires=Thu, 6 Mar 2025 213629 GMT;");` and I try to pass it into a `JsonFile` like

`FileHandlers::JsonFile cJsonFile(cookieJson); cJsonFile["Cookies"]["A1"] = A1;`

it says "terminate called after throwing an instance of 'nlohmann::json_abi_v3_11_2::detail::type_error' what():  [json.exception.type_error.302] type must be string, but is object".

The Cookie works fine, chopping and parsing are done and if I do it like this `cJsonFile["Cookies"]["A1"] = A1.dump();` and it works as well.  ChatGPT suggested to me that I should write an operator `Cookie& operator=(const nlohmann::json& other)` and pass the things trough here but this will be not triggered with `cJsonFile["Cookies"]["A1"] = A1;`.

Has anyone encountered this problem?

**Comment by nlohmann** [comment]
Inheriting from nlohmann::json (as `Cookie : public nlohmann::json`) is not supported: basic_json has no virtual functions or virtual destructor, so anywhere a Cookie is used as a json/json& (e.g. through operator[], which returns basic_json&), the object is sliced to its json base and any custom Cookie behavior is bypassed. That's exactly what happens in `cJsonFile["Cookies"]["A1"] = A1;`: operator[] returns a basic_json&, so the assignment resolves to basic_json's own operator=, converting A1 via slicing -- never through a custom operator defined on Cookie (which is why the ChatGPT-suggested `Cookie::operator=(const nlohmann::json&)` never gets invoked there). The fix is composition instead of inheritance: give Cookie/JsonFile a nlohmann::json *member* (or write a proper to_json/from_json/adl_serializer for Cookie) rather than deriving from json.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
